From f5fde18b3f40e60689454d16091551007dc4fc3d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Fri, 8 Mar 2024 18:43:38 +0100
Subject: [PATCH] sushi-media-bin: Limit timeout for gst_element_get_state

When we set the timeout to GST_CLOCK_TIME_NONE, it means that we wait for the state infinitly. Limit the timeout to 100 milliseconds to avoid freeze in case the state is not received.

Closes: https://gitlab.gnome.org/GNOME/sushi/-/issues/103
---
 src/libsushi/sushi-media-bin.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/libsushi/sushi-media-bin.c b/src/libsushi/sushi-media-bin.c
index 585db5d..78b2232 100644
--- a/src/libsushi/sushi-media-bin.c
+++ b/src/libsushi/sushi-media-bin.c
@@ -45,6 +45,8 @@
 
 #define FPS_WINDOW_SIZE          2  /* Window size in seconds to calculate fps */
 
+#define GET_STATE_TIMEOUT        GST_SECOND / G_GINT64_CONSTANT (10)
+
 #define SMB_ICON_SIZE            GTK_ICON_SIZE_BUTTON
 
 #define SMB_ICON_NAME_PLAY       "media-playback-start-symbolic"
@@ -773,7 +775,7 @@ sushi_media_bin_fullscreen_apply (SushiMediaBin *self, gboolean fullscreen)
 
       /* Init new pipeline */
       gst_element_set_state (priv->play, GST_STATE_PAUSED);
-      gst_element_get_state (priv->play, NULL, NULL, GST_CLOCK_TIME_NONE);
+      gst_element_get_state (priv->play, NULL, NULL, GET_STATE_TIMEOUT);
 
       /* Seek to position */
       gst_element_seek_simple (priv->play, GST_FORMAT_TIME,
@@ -785,7 +787,7 @@ sushi_media_bin_fullscreen_apply (SushiMediaBin *self, gboolean fullscreen)
       if (priv->state == GST_STATE_PLAYING)
         {
           gst_element_set_state (priv->play, GST_STATE_PLAYING);
-          gst_element_get_state (priv->play, NULL, NULL, GST_CLOCK_TIME_NONE);
+          gst_element_get_state (priv->play, NULL, NULL, GET_STATE_TIMEOUT);
         }
     }
 
-- 
GitLab

